From 1a2acf6afcd60a027f6a1b7e564a95255bcee69d Mon Sep 17 00:00:00 2001
From: linmin <linmin@eswincomputing.com>
Date: Mon, 13 May 2024 14:15:17 +0800
Subject: [PATCH 022/109] WIN2030-15099:feat(web-server):Add eeprom callback to
 web-server

Changelogs:
1.Call the eeprom API in web-server callback, like save_sys_username_password()...etc.

Change-Id: I05db0770e10d1e9e1bcd7aa017e176a19d7abee0
---
 Core/Inc/hf_common.h  |   5 +-
 Core/Src/hf_common.c  |  33 +++++++++++--
 Core/Src/web-server.c | 112 +++++++++++++++++++++++++++++++-----------
 3 files changed, 115 insertions(+), 35 deletions(-)

diff --git a/Core/Inc/hf_common.h b/Core/Inc/hf_common.h
index 17535c0..8519e54 100644
--- a/Core/Inc/hf_common.h
+++ b/Core/Inc/hf_common.h
@@ -152,6 +152,7 @@ typedef struct {
 #define ES_EEPROM_INFO_TEST 0
 
 /* functions prototypes ---------------------------------------------*/
+void hexstr2mac(uint8_t *mac, char *hexstr);
 
 void hf_http_task(void *argument);
 void es_eeprom_wp(uint8_t flag);
@@ -174,8 +175,8 @@ int es_set_mcu_netmask(uint8_t *p_netmask_address);
 int es_get_mcu_gateway(uint8_t *p_gateway_address);
 int es_set_mcu_gateway(uint8_t *p_gateway_address);
 
-int es_get_admin_info(char *p_admin_name, char *p_admin_password);
-int es_set_admin_info(char *p_admin_name, char *p_admin_password);
+int es_get_username_password(char *p_admin_name, char *p_admin_password);
+int es_set_username_password(const char *p_admin_name, const char *p_admin_password);
 
 int is_som_pwr_lost_resume(void);
 int es_set_som_pwr_lost_resume_attr(int isResumePwrLost);
diff --git a/Core/Src/hf_common.c b/Core/Src/hf_common.c
index 0bcbc87..080600f 100644
--- a/Core/Src/hf_common.c
+++ b/Core/Src/hf_common.c
@@ -113,6 +113,32 @@ static u32_t ntohl_seq(uint8_t *p_nlmask)
 	return hlmask;
 }
 
+static int hex2num(char c)
+{
+	if (c >= '0' && c <= '9')
+		return c - '0';
+	if (c >= 'a' && c <= 'f')
+		return c - 'a' + 10;
+	if (c >= 'A' && c <= 'F')
+		return c - 'A' + 10;
+	return -1;
+}
+
+void hexstr2mac(uint8_t *mac, char *hexstr)
+{
+	int i = 0;
+
+	while (i < 6) {
+		if (' ' == *hexstr || ':' == *hexstr || '"' == *hexstr || '\'' == *hexstr) {
+			hexstr++;
+			continue;
+		}
+		*(mac + i) = (hex2num(*hexstr) <<4) | hex2num(*(hexstr + 1));
+		i++;
+		hexstr += 2;
+	}
+}
+#if 0
 int _write(int fd, char *ch, int len)
 {
 	uint8_t val = '\r';
@@ -125,6 +151,7 @@ int _write(int fd, char *ch, int len)
 	}
 	return length;
 }
+#endif
 
 
 void es_eeprom_wp(uint8_t flag)
@@ -522,7 +549,7 @@ int es_set_mcu_gateway(uint8_t *p_gateway_address)
 }
 
 // get and set admin info
-int es_get_admin_info(char *p_admin_name, char *p_admin_password)
+int es_get_username_password(char *p_admin_name, char *p_admin_password)
 {
 	if (NULL == p_admin_name)
 		return -1;
@@ -538,7 +565,7 @@ int es_get_admin_info(char *p_admin_name, char *p_admin_password)
 	return 0;
 }
 
-int es_set_admin_info(char *p_admin_name, char *p_admin_password)
+int es_set_username_password(const char *p_admin_name, const char *p_admin_password)
 {
 	if (NULL == p_admin_name)
 		return -1;
@@ -806,7 +833,7 @@ int es_eeprom_info_test(void)
 	eeprom_debug("Get magicNumber:0x%lx\n", carrier_board_info.magicNumber);
 
 	/* admin info test */
-	es_get_admin_info(admin_name, admin_password);
+	es_get_username_password(admin_name, admin_password);
 	eeprom_debug("Get admin_name:%s, admin_password:%s\n", admin_name, admin_password);
 
 	/* mac test */
diff --git a/Core/Src/web-server.c b/Core/Src/web-server.c
index 7c46bab..a54ec7f 100644
--- a/Core/Src/web-server.c
+++ b/Core/Src/web-server.c
@@ -10412,12 +10412,13 @@ void parseCredentials(const char *readBuffer, char *username, char *password) {
 //保存用户设置的用户名密码 0 succ,1 fail
 int save_sys_username_password(const char *username, const char *password){
 	assert(username!=NULL&&password!=NULL);
-    printf("TODO call save_sys_username_password  \n");
-	return 0;
+
+	return es_set_username_password(username, password);
 }
 //查询用户设置的用户名密码 0 succ,1 fail
 void get_sys_username_password( char *username,  char *password){
-    printf("TODO call get_sys_username_password \n");
+
+	es_get_username_password(username, password);
 	return ;
 }
 
@@ -10589,22 +10590,64 @@ typedef struct  {
 
 
 NETInfo get_net_info() {
-	printf("TODO call get_net_info\n");
+	NETInfo example;
+	ip4_addr_t ip4_addr;
+	uint8_t buf[4];
+	char *p_addr;
+	uint8_t mac[6];
+
+	/* get ipaddr */
+	es_get_mcu_ipaddr(buf);
+	IP4_ADDR(&ip4_addr, buf[0], buf[1], buf[2], buf[3]);
+	p_addr = ip4addr_ntoa(&ip4_addr);
+	strcpy(example.ipaddr, p_addr);
+
+	/* get netmask */
+	es_get_mcu_netmask(buf);
+	IP4_ADDR(&ip4_addr, buf[0], buf[1], buf[2], buf[3]);
+	p_addr = ip4addr_ntoa(&ip4_addr);
+	strcpy(example.subnetwork, p_addr);
+
+	/* get gateway */
+	es_get_mcu_gateway(buf);
+	IP4_ADDR(&ip4_addr, buf[0], buf[1], buf[2], buf[3]);
+	p_addr = ip4addr_ntoa(&ip4_addr);
+	strcpy(example.gateway, p_addr);
+
+	es_get_mcu_mac(mac);
+	memset(example.macaddr, 0, sizeof(example.macaddr));
+	snprintf(example.macaddr, sizeof(example.macaddr), "%02x.%02x.%02x.%02x.%02x.%02x",
+			mac[0],mac[1],mac[2],mac[3],mac[4],mac[5]);
 
-	NETInfo example = {
-        "192.168.1.1",
-        "01:23:45:67:89:AB",
-        "255.255.255.0",
-        "192.168.1.254"
-    };
-    return example;
+	return example;
 }
 
 
 int set_net_info(NETInfo netinfo) {
-	printf("TODO call set_net_info\n");
+	u32_t naddr;
+	u32_t haddr;
+	uint8_t mac[6];
 
-    return 0;
+	/* set ipaddr */
+	naddr = ipaddr_addr(netinfo.ipaddr);
+	haddr = PP_NTOHL(naddr);
+	es_set_mcu_ipaddr((uint8_t *)&haddr);
+
+	/* set netmask */
+	naddr = ipaddr_addr(netinfo.subnetwork);
+	haddr = PP_NTOHL(naddr);
+	es_set_mcu_netmask((uint8_t *)&haddr);
+
+	/* set gateway */
+	naddr = ipaddr_addr(netinfo.gateway);
+	haddr = PP_NTOHL(naddr);
+	es_set_mcu_gateway((uint8_t *)&haddr);
+
+	/* set mac */
+	hexstr2mac(mac, netinfo.macaddr);
+	es_set_mcu_mac(mac);
+
+	return 0;
 }
 
 
@@ -10640,17 +10683,24 @@ typedef struct {
 } DIPSwitchInfo;
 
 DIPSwitchInfo get_dip_switch(){
-	printf("TODO call get_dip_switch\n");
+	uint8_t som_dip_switch_state;
 	DIPSwitchInfo dipSwitchInfo;
-	dipSwitchInfo.dip01 = 0;
-	dipSwitchInfo.dip02 = 0;
-	dipSwitchInfo.dip03 = 1;
-	dipSwitchInfo.dip04 = 1;
-    return dipSwitchInfo;
+
+	es_get_som_dip_switch_soft_state(&som_dip_switch_state);
+	dipSwitchInfo.dip01 = 0x1 & som_dip_switch_state;
+	dipSwitchInfo.dip02 = (0x2 & som_dip_switch_state) >> 1;
+	dipSwitchInfo.dip03 = (0x4 & som_dip_switch_state) >> 2;
+	dipSwitchInfo.dip04 = (0x8 & som_dip_switch_state) >> 3;
+
+	return dipSwitchInfo;
 }
 int set_dip_switch(DIPSwitchInfo dipSwitchInfo){
-	printf("TODO call set_dip_switch\n");
-    return 0;//0:success ,other:error
+	uint8_t som_dip_switch_state;
+
+	som_dip_switch_state =  ((0x1 & dipSwitchInfo.dip04) << 3) | ((0x1 & dipSwitchInfo.dip03) << 2)
+				| ((0x1 & dipSwitchInfo.dip02) << 1)| (0x1 & dipSwitchInfo.dip01);
+
+	return es_set_som_dip_switch_soft_state(som_dip_switch_state);
 }
 
 typedef struct {
@@ -10702,15 +10752,17 @@ typedef struct {
 } CBSimpleInfo;//carrie board
 
 CBSimpleInfo get_cb_info(){
-	printf("TODO call get_cb_info\n");
-	CBSimpleInfo example= {
-        0,
-        1,
-        2,
-        3,
-		"v1.10.1"
-    };
-    return example;
+	CarrierBoardInfo carrierBoardInfo;
+	CBSimpleInfo example;
+
+	es_get_carrier_borad_info(&carrierBoardInfo);
+	example.magicNumber = carrierBoardInfo.magicNumber;
+	example.formatVersionNumber = carrierBoardInfo.formatVersionNumber;
+	example.productIdentifier = carrierBoardInfo.productIdentifier;
+	example.pcbRevision = carrierBoardInfo.pcbRevision;
+	memcpy(example.boardSerialNumber, carrierBoardInfo.boardSerialNumber, sizeof(example.boardSerialNumber));
+
+	return example;
 }
 typedef struct {
 	int year;
-- 
2.25.1

